File: apps/mobile/src/services/attribution_service.ts
import { ApiClient } from './api_client';
import type {
  RecipeAttribution,
  ContributorProfile,
  CommunityMetricsData,
  MetricsTimeframe,
  ContributorAchievement,
  RecipeAttributionResponse,
  ContributorProfileResponse,
  CommunityMetricsResponse,
  AchievementCreationResponse,
} from '@imkitchen/shared-types';

export class AttributionService {
  private apiClient: ApiClient;

  constructor() {
    this.apiClient = new ApiClient();
  }

  /**
   * Get recipe attribution information
   */
  async getRecipeAttribution(recipeId: string): Promise<RecipeAttribution> {
    try {
      const response = await this.apiClient.get<RecipeAttributionResponse>(
        `/api/v1/recipes/${recipeId}/attribution`
      );

      return response.attribution;
    } catch (error) {
      console.error('Failed to get recipe attribution:', error);
      throw new Error('Failed to get recipe attribution');
    }
  }

  /**
   * Get contributor profile information
   */
  async getContributorProfile(contributorId: string): Promise<ContributorProfile> {
    try {
      const response = await this.apiClient.get<ContributorProfileResponse>(
        `/api/v1/contributors/${contributorId}/profile`
      );

      return response.profile;
    } catch (error) {
      console.error('Failed to get contributor profile:', error);
      throw new Error('Failed to get contributor profile');
    }
  }

  /**
   * Get recipe metrics for a specific recipe
   */
  async getRecipeMetrics(
    recipeId: string,
    timeframe: MetricsTimeframe = 'week'
  ): Promise<CommunityMetricsData> {
    try {
      const response = await this.apiClient.get<CommunityMetricsResponse>(
        `/api/v1/recipes/${recipeId}/metrics?timeframe=${timeframe}`
      );

      return response.metrics;
    } catch (error) {
      console.error('Failed to get recipe metrics:', error);
      throw new Error('Failed to get recipe metrics');
    }
  }

  /**
   * Get contributor metrics
   */
  async getContributorMetrics(
    contributorId: string,
    timeframe: MetricsTimeframe = 'week'
  ): Promise<CommunityMetricsData> {
    try {
      const response = await this.apiClient.get<CommunityMetricsResponse>(
        `/api/v1/contributors/${contributorId}/metrics?timeframe=${timeframe}`
      );

      return response.metrics;
    } catch (error) {
      console.error('Failed to get contributor metrics:', error);
      throw new Error('Failed to get contributor metrics');
    }
  }

  /**
   * Get personal metrics for the current user
   */
  async getPersonalMetrics(timeframe: MetricsTimeframe = 'week'): Promise<CommunityMetricsData> {
    try {
      const response = await this.apiClient.get<CommunityMetricsResponse>(
        `/api/v1/user/metrics?timeframe=${timeframe}`
      );

      return response.metrics;
    } catch (error) {
      console.error('Failed to get personal metrics:', error);
      throw new Error('Failed to get personal metrics');
    }
  }

  /**
   * Get community overview metrics
   */
  async getCommunityOverviewMetrics(timeframe: MetricsTimeframe = 'week'): Promise<CommunityMetricsData> {
    try {
      const response = await this.apiClient.get<CommunityMetricsResponse>(
        `/api/v1/community/metrics?timeframe=${timeframe}`
      );

      return response.metrics;
    } catch (error) {
      console.error('Failed to get community overview metrics:', error);
      throw new Error('Failed to get community overview metrics');
    }
  }

  /**
   * Get contributor achievements
   */
  async getContributorAchievements(contributorId: string): Promise<ContributorAchievement[]> {
    try {
      const response = await this.apiClient.get<{ achievements: ContributorAchievement[] }>(
        `/api/v1/contributors/${contributorId}/achievements`
      );

      return response.achievements;
    } catch (error) {
      console.error('Failed to get contributor achievements:', error);
      throw new Error('Failed to get contributor achievements');
    }
  }

  /**
   * Get user's own achievements
   */
  async getPersonalAchievements(): Promise<ContributorAchievement[]> {
    try {
      const response = await this.apiClient.get<{ achievements: ContributorAchievement[] }>(
        '/api/v1/user/achievements'
      );

      return response.achievements;
    } catch (error) {
      console.error('Failed to get personal achievements:', error);
      throw new Error('Failed to get personal achievements');
    }
  }

  /**
   * Update attribution preferences
   */
  async updateAttributionPreferences(preferences: {
    preserveAttribution: boolean;
    allowDerivatives: boolean;
    requireNotification: boolean;
  }): Promise<void> {
    try {
      await this.apiClient.put('/api/v1/user/attribution-preferences', preferences);
    } catch (error) {
      console.error('Failed to update attribution preferences:', error);
      throw new Error('Failed to update attribution preferences');
    }
  }

  /**
   * Report attribution issue
   */
  async reportAttributionIssue(
    recipeId: string,
    issueType: 'missing_attribution' | 'incorrect_attribution' | 'unauthorized_use',
    description: string
  ): Promise<void> {
    try {
      await this.apiClient.post('/api/v1/attribution/report', {
        recipe_id: recipeId,
        issue_type: issueType,
        description,
      });
    } catch (error) {
      console.error('Failed to report attribution issue:', error);
      throw new Error('Failed to report attribution issue');
    }
  }

  /**
   * Get recipe chain (provenance tracking)
   */
  async getRecipeChain(recipeId: string): Promise<{
    chain: Array<{
      contributorId: string;
      contributorName: string;
      adaptedAt: Date;
      recipeId: string;
    }>;
    totalAdaptations: number;
  }> {
    try {
      const response = await this.apiClient.get<{
        chain: Array<{
          contributor_id: string;
          contributor_name: string;
          adapted_at: string;
          recipe_id: string;
        }>;
        total_adaptations: number;
      }>(`/api/v1/recipes/${recipeId}/chain`);

      return {
        chain: response.chain.map(link => ({
          contributorId: link.contributor_id,
          contributorName: link.contributor_name,
          adaptedAt: new Date(link.adapted_at),
          recipeId: link.recipe_id,
        })),
        totalAdaptations: response.total_adaptations,
      };
    } catch (error) {
      console.error('Failed to get recipe chain:', error);
      throw new Error('Failed to get recipe chain');
    }
  }

  /**
   * Get trending contributors
   */
  async getTrendingContributors(
    limit: number = 10,
    timeframe: MetricsTimeframe = 'week'
  ): Promise<Array<{
    contributor: ContributorProfile;
    metrics: {
      newRecipes: number;
      totalImports: number;
      averageRating: number;
      trendingScore: number;
    };
  }>> {
    try {
      const response = await this.apiClient.get<{
        contributors: Array<{
          contributor: ContributorProfile;
          metrics: {
            new_recipes: number;
            total_imports: number;
            average_rating: number;
            trending_score: number;
          };
        }>;
      }>(`/api/v1/contributors/trending?limit=${limit}&timeframe=${timeframe}`);

      return response.contributors.map(item => ({
        contributor: item.contributor,
        metrics: {
          newRecipes: item.metrics.new_recipes,
          totalImports: item.metrics.total_imports,
          averageRating: item.metrics.average_rating,
          trendingScore: item.metrics.trending_score,
        },
      }));
    } catch (error) {
      console.error('Failed to get trending contributors:', error);
      throw new Error('Failed to get trending contributors');
    }
  }

  /**
   * Get contributor leaderboard
   */
  async getContributorLeaderboard(
    category: 'recipes' | 'imports' | 'ratings' | 'achievements',
    timeframe: MetricsTimeframe = 'month',
    limit: number = 50
  ): Promise<Array<{
    rank: number;
    contributor: ContributorProfile;
    score: number;
    change: number; // Position change from previous period
  }>> {
    try {
      const response = await this.apiClient.get<{
        leaderboard: Array<{
          rank: number;
          contributor: ContributorProfile;
          score: number;
          change: number;
        }>;
      }>(`/api/v1/contributors/leaderboard?category=${category}&timeframe=${timeframe}&limit=${limit}`);

      return response.leaderboard;
    } catch (error) {
      console.error('Failed to get contributor leaderboard:', error);
      throw new Error('Failed to get contributor leaderboard');
    }
  }

  /**
   * Award achievement to a contributor (admin only)
   */
  async awardAchievement(
    contributorId: string,
    achievementType: string,
    metadata?: Record<string, any>
  ): Promise<ContributorAchievement> {
    try {
      const response = await this.apiClient.post<AchievementCreationResponse>(
        `/api/v1/contributors/${contributorId}/achievements`,
        {
          achievement_type: achievementType,
          metadata,
        }
      );

      return response.achievement;
    } catch (error) {
      console.error('Failed to award achievement:', error);
      throw new Error('Failed to award achievement');
    }
  }

  /**
   * Get detailed engagement analytics
   */
  async getEngagementAnalytics(
    recipeId: string,
    timeframe: MetricsTimeframe = 'week'
  ): Promise<{
    dailyViews: Array<{ date: Date; views: number }>;
    referralSources: Array<{ source: string; count: number }>;
    userActions: Array<{ action: string; count: number }>;
    geographicData: Array<{ country: string; count: number }>;
  }> {
    try {
      const response = await this.apiClient.get<{
        daily_views: Array<{ date: string; views: number }>;
        referral_sources: Array<{ source: string; count: number }>;
        user_actions: Array<{ action: string; count: number }>;
        geographic_data: Array<{ country: string; count: number }>;
      }>(`/api/v1/recipes/${recipeId}/analytics?timeframe=${timeframe}`);

      return {
        dailyViews: response.daily_views.map(item => ({
          date: new Date(item.date),
          views: item.views,
        })),
        referralSources: response.referral_sources,
        userActions: response.user_actions,
        geographicData: response.geographic_data,
      };
    } catch (error) {
      console.error('Failed to get engagement analytics:', error);
      throw new Error('Failed to get engagement analytics');
    }
  }

  /**
   * Export attribution data for GDPR compliance
   */
  async exportAttributionData(): Promise<{
    attributions: RecipeAttribution[];
    contributions: Array<{
      recipeId: string;
      title: string;
      createdAt: Date;
      totalImports: number;
    }>;
    achievements: ContributorAchievement[];
  }> {
    try {
      const response = await this.apiClient.get<{
        attributions: RecipeAttribution[];
        contributions: Array<{
          recipe_id: string;
          title: string;
          created_at: string;
          total_imports: number;
        }>;
        achievements: ContributorAchievement[];
      }>('/api/v1/user/attribution-export');

      return {
        attributions: response.attributions,
        contributions: response.contributions.map(item => ({
          recipeId: item.recipe_id,
          title: item.title,
          createdAt: new Date(item.created_at),
          totalImports: item.total_imports,
        })),
        achievements: response.achievements,
      };
    } catch (error) {
      console.error('Failed to export attribution data:', error);
      throw new Error('Failed to export attribution data');
    }
  }
}

File: apps/backend/internal/services/attribution_service.go
package services

import (
	"context"
	"fmt"
	"time"

	"github.com/google/uuid"
	"github.com/imkitchen/backend/internal/models"
	"github.com/imkitchen/backend/internal/repositories"
)

type AttributionService struct {
	attributionRepo repositories.AttributionRepository
	userRepo        repositories.UserRepository
	recipeRepo      repositories.RecipeRepository
	metricsRepo     repositories.MetricsRepository
}

func NewAttributionService(
	attributionRepo repositories.AttributionRepository,
	userRepo repositories.UserRepository,
	recipeRepo repositories.RecipeRepository,
	metricsRepo repositories.MetricsRepository,
) *AttributionService {
	return &AttributionService{
		attributionRepo: attributionRepo,
		userRepo:        userRepo,
		recipeRepo:      recipeRepo,
		metricsRepo:     metricsRepo,
	}
}

// RecipeAttribution type alias for consistency with models package
type RecipeAttribution = models.RecipeAttribution

// Type aliases for consistency with models package
type CommunityMetrics = models.CommunityMetrics
type RecipeChainLink = models.RecipeChainLink
type EngagementStats = models.EngagementStats
type ImportCustomizations = models.ImportCustomizations

// Type aliases for repository types
type ContributorBadge = repositories.ContributorBadge
type ContributorAchievement = repositories.ContributorAchievement

// Type aliases for metrics
type MetricsOverview = repositories.RecipeOverview
type PopularityMetrics = repositories.RecipePopularity
type EngagementMetrics = repositories.RecipeEngagement
type GeographicMetrics = repositories.RecipeGeographic

type ContributorProfile struct {
	ID            string                   `json:"id"`
	Username      string                   `json:"username"`
	DisplayName   string                   `json:"display_name"`
	AvatarURL     *string                  `json:"avatar_url"`
	TotalRecipes  int                      `json:"total_recipes"`
	AverageRating float64                  `json:"average_rating"`
	TotalImports  int                      `json:"total_imports"`
	JoinedAt      time.Time                `json:"joined_at"`
	Badges        []ContributorBadge       `json:"badges"`
	Achievements  []ContributorAchievement `json:"achievements"`
	Bio           *string                  `json:"bio"`
	Location      *string                  `json:"location"`
	Website       *string                  `json:"website"`
}


type CommunityMetricsData struct {
	Overview     *MetricsOverview     `json:"overview,omitempty"`
	Popularity   *PopularityMetrics   `json:"popularity,omitempty"`
	Engagement   *EngagementMetrics   `json:"engagement,omitempty"`
	Geographic   *GeographicMetrics   `json:"geographic,omitempty"`
	Achievements []ContributorAchievement `json:"achievements,omitempty"`
}


type RegionMetric struct {
	Name  string `json:"name"`
	Flag  string `json:"flag"`
	Count int    `json:"count"`
}

type MetricsTimeframe string

const (
	TimeframeDay     MetricsTimeframe = "day"
	TimeframeWeek    MetricsTimeframe = "week"
	TimeframeMonth   MetricsTimeframe = "month"
	TimeframeQuarter MetricsTimeframe = "quarter"
	TimeframeYear    MetricsTimeframe = "year"
	TimeframeAll     MetricsTimeframe = "all"
)

func (s *AttributionService) CreateRecipeAttribution(ctx context.Context, req *models.RecipeImportRequest, personalRecipeID string) (*RecipeAttribution, error) {
	// Convert string ID to UUID
	recipeUUID, err := uuid.Parse(req.CommunityRecipeID)
	if err != nil {
		return nil, fmt.Errorf("invalid recipe ID format: %w", err)
	}

	// Get the original recipe details
	originalRecipe, err := s.recipeRepo.GetCommunityRecipeByID(ctx, recipeUUID)
	if err != nil {
		return nil, fmt.Errorf("failed to get original recipe: %w", err)
	}

	// Get contributor details
	contributor, err := s.userRepo.GetByID(originalRecipe.UserID)
	if err != nil {
		return nil, fmt.Errorf("failed to get contributor: %w", err)
	}

	// Create attribution record
	attribution := &models.RecipeAttribution{
		RecipeID:              personalRecipeID,
		OriginalContributorID: originalRecipe.UserID.String(),
		OriginalContributor:   contributor.Username,
		ImportDate:            time.Now(),
		PreserveAttribution:   req.PreserveAttribution,
		Customizations:        extractRecipeImportCustomizations(req.Customizations),
	}

	if err := s.attributionRepo.Create(ctx, attribution); err != nil {
		return nil, fmt.Errorf("failed to create attribution record: %w", err)
	}

	// Build recipe chain
	chain, err := s.buildRecipeChain(ctx, req.CommunityRecipeID)
	if err != nil {
		// Log warning but don't fail the request
		fmt.Printf("Warning: failed to build recipe chain: %v\n", err)
	}

	// Get community metrics
	communityMetrics, err := s.getCommunityMetrics(ctx, req.CommunityRecipeID)
	if err != nil {
		// Log warning but don't fail the request
		fmt.Printf("Warning: failed to get community metrics: %v\n", err)
		communityMetrics = &CommunityMetrics{}
	}

	// Increment import count
	go func() {
		if err := s.attributionRepo.IncrementImportCount(context.Background(), req.CommunityRecipeID); err != nil {
			fmt.Printf("Warning: failed to increment import count: %v\n", err)
		}
	}()

	return &RecipeAttribution{
		ID:                    attribution.ID,
		RecipeID:              personalRecipeID,
		OriginalContributorID: originalRecipe.UserID.String(),
		OriginalContributor:   contributor.Username,
		ImportDate:            attribution.ImportDate,
		PreserveAttribution:   attribution.PreserveAttribution,
		Customizations:        attribution.Customizations,
		CommunityMetrics:      *communityMetrics,
		RecipeChain:          chain,
	}, nil
}

func (s *AttributionService) GetRecipeAttribution(ctx context.Context, recipeID string) (*RecipeAttribution, error) {
	attribution, err := s.attributionRepo.GetByRecipeID(ctx, recipeID)
	if err != nil {
		return nil, fmt.Errorf("failed to get attribution: %w", err)
	}

	// Convert string ID to UUID
	contributorUUID, err := uuid.Parse(attribution.OriginalContributorID)
	if err != nil {
		return nil, fmt.Errorf("invalid contributor ID format: %w", err)
	}

	// Get contributor details
	contributor, err := s.userRepo.GetByID(contributorUUID)
	if err != nil {
		return nil, fmt.Errorf("failed to get contributor: %w", err)
	}

	// Build recipe chain
	chain, err := s.buildRecipeChain(ctx, attribution.RecipeID)
	if err != nil {
		fmt.Printf("Warning: failed to build recipe chain: %v\n", err)
	}

	// Get community metrics
	communityMetrics, err := s.getCommunityMetrics(ctx, attribution.RecipeID)
	if err != nil {
		fmt.Printf("Warning: failed to get community metrics: %v\n", err)
		communityMetrics = &CommunityMetrics{}
	}

	// Get engagement stats
	engagementStats, err := s.getEngagementStats(ctx, recipeID)
	if err != nil {
		fmt.Printf("Warning: failed to get engagement stats: %v\n", err)
	}

	return &RecipeAttribution{
		ID:                    attribution.ID,
		RecipeID:              recipeID,
		OriginalContributorID: attribution.OriginalContributorID,
		OriginalContributor:   contributor.Username,
		ImportDate:            attribution.ImportDate,
		PreserveAttribution:   attribution.PreserveAttribution,
		Customizations:        attribution.Customizations,
		CommunityMetrics:      *communityMetrics,
		RecipeChain:          chain,
		EngagementStats:      engagementStats,
	}, nil
}

func (s *AttributionService) GetContributorProfile(ctx context.Context, contributorID string) (*ContributorProfile, error) {
	// Convert string ID to UUID
	userUUID, err := uuid.Parse(contributorID)
	if err != nil {
		return nil, fmt.Errorf("invalid contributor ID format: %w", err)
	}
	
	user, err := s.userRepo.GetByID(userUUID)
	if err != nil {
		return nil, fmt.Errorf("failed to get user: %w", err)
	}

	// Get contributor stats
	stats, err := s.attributionRepo.GetContributorStats(ctx, contributorID)
	if err != nil {
		return nil, fmt.Errorf("failed to get contributor stats: %w", err)
	}

	// Get badges and achievements
	badges, err := s.attributionRepo.GetContributorBadges(ctx, contributorID)
	if err != nil {
		fmt.Printf("Warning: failed to get contributor badges: %v\n", err)
		badges = []ContributorBadge{}
	}

	achievements, err := s.attributionRepo.GetContributorAchievements(ctx, contributorID)
	if err != nil {
		fmt.Printf("Warning: failed to get contributor achievements: %v\n", err)
		achievements = []ContributorAchievement{}
	}

	avatarURL := &user.AvatarURL
	
	return &ContributorProfile{
		ID:            user.ID.String(),
		Username:      user.Username,
		DisplayName:   user.DisplayName,
		AvatarURL:     avatarURL,
		TotalRecipes:  stats.TotalRecipes,
		AverageRating: stats.AverageRating,
		TotalImports:  stats.TotalImports,
		JoinedAt:      user.CreatedAt,
		Badges:        badges,
		Achievements:  achievements,
		Bio:           user.Bio,
		Location:      user.Location,
		Website:       user.Website,
	}, nil
}

func (s *AttributionService) GetRecipeMetrics(ctx context.Context, recipeID string, timeframe MetricsTimeframe) (*CommunityMetricsData, error) {
	// Get basic metrics
	overview, err := s.metricsRepo.GetRecipeOverview(ctx, recipeID, string(timeframe))
	if err != nil {
		return nil, fmt.Errorf("failed to get recipe overview: %w", err)
	}

	// Get popularity metrics
	popularity, err := s.metricsRepo.GetRecipePopularity(ctx, recipeID, string(timeframe))
	if err != nil {
		fmt.Printf("Warning: failed to get popularity metrics: %v\n", err)
	}

	// Get engagement metrics
	engagement, err := s.metricsRepo.GetRecipeEngagement(ctx, recipeID, string(timeframe))
	if err != nil {
		fmt.Printf("Warning: failed to get engagement metrics: %v\n", err)
	}

	// Get geographic metrics
	geographic, err := s.metricsRepo.GetRecipeGeographic(ctx, recipeID, string(timeframe))
	if err != nil {
		fmt.Printf("Warning: failed to get geographic metrics: %v\n", err)
	}

	return &CommunityMetricsData{
		Overview:   overview,
		Popularity: popularity,
		Engagement: engagement,
		Geographic: geographic,
	}, nil
}

func (s *AttributionService) GetContributorMetrics(ctx context.Context, contributorID string, timeframe MetricsTimeframe) (*CommunityMetricsData, error) {
	// Get contributor overview
	overview, err := s.metricsRepo.GetContributorOverview(ctx, contributorID, string(timeframe))
	if err != nil {
		return nil, fmt.Errorf("failed to get contributor overview: %w", err)
	}

	// Get achievements
	achievements, err := s.attributionRepo.GetContributorAchievements(ctx, contributorID)
	if err != nil {
		fmt.Printf("Warning: failed to get achievements: %v\n", err)
	}

	// Convert contributor overview to metrics overview format
	metricsOverview := &MetricsOverview{
		TotalViews:   overview.TotalViews,
		TotalImports: overview.TotalImports,
		TotalRatings: overview.TotalRecipes,
		AvgRating:    overview.AverageRating,
	}
	
	return &CommunityMetricsData{
		Overview:     metricsOverview,
		Achievements: achievements,
	}, nil
}

func (s *AttributionService) AwardAchievement(ctx context.Context, contributorID, achievementType string, metadata map[string]interface{}) error {
	achievement := s.createAchievement(achievementType, metadata)
	if achievement == nil {
		return fmt.Errorf("unknown achievement type: %s", achievementType)
	}

	return s.attributionRepo.CreateAchievement(ctx, contributorID, achievement)
}

// Helper methods

func (s *AttributionService) buildRecipeChain(ctx context.Context, recipeID string) ([]RecipeChainLink, error) {
	chain, err := s.attributionRepo.GetRecipeChain(ctx, recipeID)
	if err != nil {
		return nil, fmt.Errorf("failed to get recipe chain: %w", err)
	}

	var links []RecipeChainLink
	for _, link := range chain {
		// Convert string ID to UUID
		contributorUUID, err := uuid.Parse(link.ContributorID)
		if err != nil {
			fmt.Printf("Warning: invalid contributor ID %s: %v\n", link.ContributorID, err)
			continue
		}
		
		contributor, err := s.userRepo.GetByID(contributorUUID)
		if err != nil {
			fmt.Printf("Warning: failed to get contributor %s: %v\n", link.ContributorID, err)
			continue
		}

		links = append(links, RecipeChainLink{
			ContributorID:   link.ContributorID,
			ContributorName: contributor.Username,
			Contribution:    link.Contribution,
			Timestamp:       link.Timestamp,
		})
	}

	return links, nil
}

func (s *AttributionService) getCommunityMetrics(ctx context.Context, recipeID string) (*CommunityMetrics, error) {
	// TODO: Implement full community metrics retrieval
	// For now, return default values
	return &CommunityMetrics{
		TotalImports:  0,
		AverageRating: 0.0,
		TotalRatings:  0,
		TrendingScore: 0.0,
	}, nil
}

func (s *AttributionService) getEngagementStats(ctx context.Context, recipeID string) (*EngagementStats, error) {
	// TODO: Implement full engagement stats retrieval
	// For now, return default values
	return &EngagementStats{
		Views:    0,
		Likes:    0,
		Saves:    0,
		Shares:   0,
		Comments: 0,
	}, nil
}

func (s *AttributionService) createAchievement(achievementType string, metadata map[string]interface{}) *ContributorAchievement {
	achievements := map[string]ContributorAchievement{
		"first_recipe": {
			Name:        "First Recipe",
			Description: "Shared your first recipe with the community",
			Category:    "milestone",
			Progress:    1,
			MaxProgress: 1,
		},
		"popular_recipe": {
			Name:        "Popular Recipe",
			Description: "One of your recipes reached 100+ imports",
			Category:    "popularity", 
			Progress:    1,
			MaxProgress: 1,
		},
		"community_favorite": {
			Name:        "Community Favorite",
			Description: "Recipe voted as community favorite",
			Category:    "recognition",
			Progress:    1,
			MaxProgress: 1,
		},
		"prolific_creator": {
			Name:        "Prolific Creator", 
			Description: "Shared 50+ recipes with the community",
			Category:    "milestone",
			Progress:    50,
			MaxProgress: 50,
		},
		"trending_chef": {
			Name:        "Trending Chef",
			Description: "Had a recipe trending for a full week",
			Category:    "trending",
			Progress:    1,
			MaxProgress: 1,
		},
	}

	achievement, exists := achievements[achievementType]
	if !exists {
		return nil
	}

	achievement.ID = fmt.Sprintf("%s_%d", achievementType, time.Now().Unix())
	unlockedAt := time.Now()
	achievement.UnlockedAt = &unlockedAt

	return &achievement
}

func extractCustomizations(customizations *models.RecipeCustomizations) []string {
	if customizations == nil {
		return []string{}
	}

	var changes []string
	if customizations.CustomTitle != nil {
		changes = append(changes, "title")
	}
	if customizations.CustomNotes != nil {
		changes = append(changes, "notes")
	}
	if customizations.ServingAdjustment != nil {
		changes = append(changes, "servings")
	}

	return changes
}

func extractRecipeImportCustomizations(customizations *models.RecipeImportCustomizations) []string {
	if customizations == nil {
		return []string{}
	}

	var changes []string
	if customizations.Title != nil {
		changes = append(changes, "title")
	}
	if customizations.Notes != nil {
		changes = append(changes, "notes")
	}
	if customizations.ServingAdjustment != nil {
		changes = append(changes, "servings")
	}

	return changes
}